src/target/riscv/riscv-011.c

```diff
@@ -455,16 +455,21 @@ static uint64_t dbus_read(struct target *target, uint16_t address)
 	dbus_status_t status;
 	uint16_t address_in;
 
-	unsigned i = 0;
-	do {
-		status = dbus_scan(target, &address_in, &value, DBUS_OP_READ, address, 0);
+	for (unsigned i = 0; i < 256; i++) {
+		dbus_scan(target, NULL, NULL, DBUS_OP_READ, address, 0);
```

Comment:
I think you should add a status = dbus_scan on the first and check that it wasn't busy. There may be no case in which it would be busy, but it's at least a sanity that may help us find bugs.

Reply:
But if it returns, that's a stale result. It doesn't tell you anything about whether the scan that we're actually doing right now is going to succeed or not.

Reply:
mmmm, this still doesn't quite make sense to me. What you really want to do is make sure that 1) the read command was seen and 2) that you got the result back. So you need to check for busy twice.

I suggest you first put a loop of `dbus_scan(DBUS_OP_READ)` where you check for busy. Then you would do  a second loop of  `dbus_scan(DBUS_OP_NOP)` until you didn't get busy (again), in which case you know you got the actual read data.

I'm checking to see if that is how v013 does this.

Reply:
yeah, confirmed that v013 has 2 separate loops.

Reply:
Though now that I think about it, in v11 NOP wasn't really a thing. So I think you could still get BUSY response even for a NOP.

Reply:
Oh just saw your comment.

BUSY = 0 just means "there was no other operation in flight when you asked me to do this one". So that means two things: 1) I am free to start the operation that you just asked me to do, otherwise I would have ignored it, and 2) the data that you read was the result from the previous operation.

By doing a read with 2 loops, you are covering the two meanings: the first one makes sure that the read goes through, and the second one makes sure that the data you saw was completed.

In your original code you were basically doing the same thing except you were using the address match as the check for case 1) and just re-sending the request over and over. Which is why I was pretty much fine with it, just thought it was confusing without a comment.

Reply:
Won't the second scan get a busy response if it happens before the read has completed? I'm thinking this is what happens:
```
scan read
Update-DR, read begins
Capture-DR, captures either busy or result
scan NOP, scanning out either busy or result
```

Reply:
Ah yeah I see what you mean. Let me go back to the v13 code to convince myself what I really meant with #39

Reply:
Argh. I had a huge write-up with fast/slow examples but accidentally closed the window.

Long story short, you need to consider that Capture-DR happens BEFORE Update-DR. The busy bit is recorded during Capture-DR, and if it is set, nothing gets sent across the DBUS/DMI during Update-DR.

So your example above is more correct as:

```
Scan READ:
  Capture-DR: busy /result captured
  Update-DR: read begins (assuming busy =0)
Scan NOP:
   Capture-DR: result captured (assuming busy = 0)
   Update-DR: NOP gets sent out
```

So yes, either of those scans can get busy responses, which is why you need to keep doing them in two separate loops. If the Scan READ reports a busy response, it won't actually send anything out over DBUS/DMI. If the Scan NOP reports a busy response, it won't actually have the right data.

Your old code worked (fine, except for the issue raised that it was kind of weird) like:
```
Scan READ:
  Capture-DIR: busy/ result & address captured
  Update-DR: Useless read (or read that is used next time) sent into the DMI
```
or
```
Scan READ:
  Capture-DR: busy/ result & (wrong) address captured
  Update-DR: Actual read to correct address sent over DMI
Scan READ:
  Capture-DR: busy/ result & address captured
  Update-DR: Useless read (or read that is used next time) sent into the DMI
```

So yes, you would only have to do one READ as long as the address matched.